### ISSUE TITLE
pynufft.forward(x) output should have a reasonable shape, probably Kd

### ISSUE THREAD

**Issue by Ritchizh** [jyhmiinlin/pynufft#1]
**#create object**
> NufftObj = pyn.pynufft(om, Nd,Kd,Jd, None)   
>  
**#this is the Forward Fourier Transform**
> data= NufftObj.forward(image)
> 
**#recalculate the density compensation**   
> NufftObj.pipe_density() 
> 
**#this is the Inverse Fourier Transform**    
> image_blur = NufftObj.adjoint(data*NufftObj.st['W'])

Uploaded "image" has the shape [256,256], "data" is the "image" spectrum which I was looking for. But somewhere in "def forward" function "data" is reshaped into [Mx1] array, where M is initialized as om.shape[0], so the Fourier spectrum can not be used or even plotted directly, which is very inconvenient.
Probably the package functions should be changed so that they do not flatten the "forward" result.

**Comment by jyhmiinlin** [comment]
Ritchizh,

Thank you for the letter.

I have revised the pynufft module. I will update the code base in a few
days.

Regards,

Jyh-Miin

2016-12-16 14:00 GMT+00:00 Ritchizh <notifications@github.com>:

**Comment by jyhmiinlin** [comment]
I updated the test_2D(), which now includes an example of spectrum. I also updated the pypi database. Please install pynufft 0.3.1.5. 

If you want to see the spectrum on the regular grid, please use y2k_DC() method.
>> y = NufftObj.forward(image) # y is the (M, ) data
>> kspectrum = NufftObj.y2k_DC(y) # convert non-uniform (M,) samples to k-spectrum (Kd array)

Please let me know if this answers your question. Thanks.

**Comment by Ritchizh** [comment]
Jyh-Miin,
Thank you very much for the update. I need some time to consider it carefully and will let you know about the result.
Kind regards,
Margarita

**Comment by Ritchizh** [comment]
Jyh-Miin,
After trying the updated version I have several questions, I would be very grateful if you clarified these moments for me:
1. **Why do you keep adjoint and inverse transforms different?** `image2 = inverse_DC(y)` is the same as `image2 = backward(st['W']*y) ` -  so is it a normalization issue, why not leave only the inverse?
The same is with "y2k_DC" and "y2k" methods - are both of them needed? What does "DC" mean?
2. **Should "om" be necessarily normalized between [-pi, pi]?**
3. What I wish to do with the help of your package:

- Experimental data are a numpy array [Nx, Ny, Nf], the forward FFT result is: `y = forward(data)`
- `kspectrum1 = y2k_DC(y)` - get the spectrum. Then I apply some transformations to it, like, for example, filtering, and receive kspectrum2.
- `y2 = k2y(kspectrum2)` - **so is it legitimate to use here "k2y"** and not some "k2y_DC"?
- finally: `image2 = inverse_DC(y2)`

For a simple image it worked fine, I will now switch to testing it on complex experimental data.

4. There is a small misprint in README, Summary, the adjoint transform, paragraph 4: should be xx2x.
The updated version is much more user-friendly, though, probably, you should also put into README all the notations and available methods, like:
x -   an image - [Nd, Nd]
xx - a scaled image - [2Nd, 2Nd]
y -  data after the forward FFT - [M, 1]
k -  a spectrum - [Kd, Kd]
-----------------------------
x2xx - scaling
xx2x - rescaling

y2k_DC (y2k) - convert data to a specrum
k2y - convert a spectrum to data  (- no "DC" here...)

xx2k - get a spectrum of an image
k2xx - get an image from a spectrum

k2k - regridding of a spectrum

------------------------------
To summarize, as I understood the logic of the package:
If you have an image/signal "x", you compute:` y = forward(x)`, and then you have three options:
1) get the k-spectrum: `kspectrum = y2k_DC(y)` (- obscuring notation)
2) do the adjoint Fourier transform:  `image2 = backward(y)` (- obscuring notation, does not convey the meaning, I would name it "adjoint")
3) do the Inverse Fourier transform:` image2 = backward(y*st['W'])` or `image2 =inverse_DC(y)`  (- obscuring notation)

Kind regards,
Margarita

**Comment by jyhmiinlin** [comment]
Margarita, 

I agree your suggestions about adjoint() and I am writing a pynufft user manual to explain the concept of NUFFT.

NUFFT and pynufft packages try to obtain data _y_ (not spectrum _k_, see the following explanations), so it requires additional inverse_DC() and y2k_DC() to recover the image and spectrum k. I would start with FFT/IFFT, then explain the NUFFT/adjoint/inverse_DC.

- **FFT**: 

There is no problem with FFT/IFFT. FFT simply transform an image to spectral data (from _x_ to _y_). _y_ is data also the _k_-spectrum. Inverse is also identical to adjoint FFT with a proper scaling (IFFT == FFT<sup>H</sup>/N). 

>>>> _y = numpy.fft.fftn(x)._ 

- **spectrum (_k_) and data (_y_) are no longer identical in pynufft**

However, NUFFT transform regular image _x_ to non-uniform data _y_. That's why forward NUFFT transform is a series of operations:

>>>> _x -> xx -> k -> y_

Here, spectrum (_k_) and data (_y_) are no longer identical in pynufft. k is the regular k-space array in the size of (Kd, Kd). y is an length-M array. 

- **Inverse is not identical to adjoint, either.**

Note scaling (x -> xx, x2xx()), interpolation (k2y()), and density compensation are the tricky parts of pynufft. 

- **Scaling and rescaling**

Scaling/Rescaling would increase the accuracy of interpolation (according to mix-max NUFFT paper). Thus scaling x2xx() always comes with the interpolation k2y(). Basically, rescaling is identical to scaling, which is also needed in adjoint() or inverse_DC(). (I will explain inverse_DC() and y2k_DC() below)

To remind users, scaling has nothing to do with sqrt(N) or N of FFT. Scaling is related to interpolation kernel.

- **Why adjoint() is different from inverse_DC()**

adjoint() is not equal to inverse, which is also different from FFT and IFFT. (in FFT, inverse FFT is equal to conjugate transpose.)

Because non-uniform coordinates induce uneven spectral k density, the spectrum is not equally sampled. Uneven density means the **adjoint() won't recover the image x**.

- **Density compensation (DC)**

Non-uniform data also cause non-uniform spectrum density. A trick called "density compensation" (DC) removes such uneven density by modulating the data y with a weighting function st['W']. That's how y2k_DC() interpolates the data back to regular k-spectrum.  

- **Why adjoint() is still needed?**

Short answer: adjoint() exists for iterative inverse problem.

Long answer: adjoint() is used in **normal equation** of iterative inverse problem. To explain the most basic idea, inverse reconstruction minimizes the cost function:

cost function = |y - Ax|<sup>2</sup>

Here A is the forward() transform of pynufft. The derivative of cost function is:

0 = A<sup>H</sup>y - A<sup>H</sup>Ax

Which is called **normal equation**. A<sup>H</sup> is the adjoint(). However, it will require a whole book to explain the details of iterative inverse problem.

Reference: Wolfram's website:
http://mathworld.wolfram.com/NormalEquation.html

- **Filtering in pynufft**

If you have the non-uniform data _y_, filtering is:

>>>>kspectrum = y2k_DC(y) # This obtains the density compensated kspectrum, no fftshift
>>>>kspectrum2 = kspectrum * your_own_filter
>>>>image=xx2x(k2xx(kspectrum2))

If image _x_ is to be filtered:

>>>>y = forward(x)
>>>>kspectrum = y2k_DC(y) # This obtains the density compensated kspectrum, no fftshift
>>>>kspectrum2 = kspectrum* your_own_filter # Filter must be Kd size
>>>>image=xx2x(k2xx(kspectrum2))

or simply ignore the scaling and interpolation:

>>>>kspectrum = xx2k(x) # neither interpolation nor scaling is requried here.
>>>>kspectrum2 = kspectrum* your_own_filter  # Filter must be Kd size
>>>>image= k2xx(kspectrum2)

- **om is normalized to [-pi and pi]**
This was the original design of Fessler's MATLAB program, and this normalization seems natural to me. The other choice is normalization to [-0.5 0.5] but I don't want to change this. 

Suppose you have an **om** normalized to [-0.5 0.5], the om2 should be multiplied by 2pi for pynufft.
>>>>om2 = om * 2 * 3.14159265359


...I will try to update the code base and documents. Thanks!

J-M

**Comment by jyhmiinlin** [comment]
I am closing this issue as transforms should be properly documented.

**Comment by Ritchizh** [comment]
Okay.
I am following your updates!
M.

**Comment by Ritchizh** [comment]
Dear Jyh-Miin,
what I really need is a Python analogue of FGG_3d_type1 Matlab function.
It is very intuitive: FGG_3d_type1(f_data, old_mesh, data size, accuracy, new_mesh). Now I am trying to implement something similar with my data and meshes and your pynufft.
M.

**Comment by jyhmiinlin** [comment]
Type 1 problem was not detailed in Fessler's paper. But it seems Type 1 is the inverse of Type 2 NUFFT. See https://web.eecs.umich.edu/~fessler/papers/files/jour/03/pre/tsp,nufft.pdf section G.

Pynufft basically implements the Type 2 NUFFT problem (forward() method). So inverse_DC() provides an approximate Type 1 NUFFT solution. However, a more accurate inverse method will require expensive conjugate gradient method, which I haven't implemented yet. 

Currently I am writing the user manual with examples, including 3D cases. So please wait.
Thanks

**Comment by Ritchizh** [comment]
Dear Jyh-Miin,
I uninstalled pynufft 3.1.8 to get the latest version 3.2.2, and had a problem with installing it via Windows cmd. When I print `pip3 install pynufft` an error occurs: 
![path_error](https://cloud.githubusercontent.com/assets/18049634/22204188/b0d63188-e181-11e6-9987-bdb30938d9ad.png)
(I think it could be due to you using Linux and writing filepaths with "/" and not "\\".)
I had to return to the version 3.1.8 - downloaded and installed  from github.
M.

**Comment by jyhmiinlin** [comment]
Thanks!
You are right. In setup.py, the ending forward slash in "package_dir={'pynufft':'./'}" caused a problem under windows. It should look like "package_dir={'pynufft':'.'}" without the ending "/".

Now pynufft-3.2.8 removes the forward slash at the end, which should work under Windows Python 3 and Linux python 3.0. Please let me know if you have question about this new version.

J-M

**Comment by Ritchizh** [comment]
Dear Jyh-Miin,
I have read the new Documentation and it is really good: nice design, clear examples and comments.

(I worked out the answer to my question :) )
M.

**Comment by jyhmiinlin** [comment]
That's great! 

J-M

**Comment by jyhmiinlin** [comment]
The documents now solve the issue.  So the issue will be closed if no further question.
